front_end/dart_tests/test_helpers.dart

```diff
@@ -0,0 +1,273 @@
+  if (asPackageReference != '') {
+    return asPackageReference;
+  }
+  // Otherwise we assume it's in web or test, and probably looks like
+  // "http://localhost:8081/foo/bar/web/main.dart"
+  var uri = Uri.parse(originalUrl);
+  return uri.path;
+}
+
+/// A class to describe all the compilation information and allow comparing it
+/// in tests.
+class CompilationContext {
+  static CompilationContext current;
+
+  var properties;
+  var module;
+  var targetName;
+  var currentFile;
+
+  static noteCurrent({properties, module, currentFile, targetName}) {
+    current = CompilationContext()
+      ..properties = properties
+      ..module = module
+      ..currentFile = currentFile
+      ..targetName = targetName;
+  }
+
+  bool equals(context) {
+    if (module != context.module) return false;
+    if (currentFile != context.currentFile) return false;
+    if (targetName != context.targetName) return false;
+    if (properties.length != context.properties.length) return false;
+    for (var i = 0; i < properties.length; i++) {
+      if (properties[i] != context.properties[i]) return false;
+    }
+    return true;
+  }
+}
+
+/// A fake Http response we can give to JavaScript to mimic the server's compiled response.
+class FakeResponse {
+  text() => futureToPromise(Future.value(_text));
+  FakeResponse(this._text);
+  String _text;
+}
+
+/// An environment, with a prefix and a List of items, but no values.
+/// Fake version of what gets returned from Dart.environments
+class Environment {
+  String prefix;
+  List<String> items;
+  Environment(this.prefix, this.items);
+}
+
+/// Create a new execution context with hard-coded fields as if we were
+/// executing a file from the web directory, so it can't just be imported with a
+/// package: URL.
+newContextInWeb() {
+  var context = FakeExecutionContext();
+  context.selectedCallFrame = allowInterop(() => FakeCallFrame.inWeb()
+    ..evaluationResult = FakeRemoteObject(value: 'foo/bar/web/main.html'));
+  return context;
+}
+
+/// Create a new execution context with hard-coded fields as if we were
+/// executing a file from the lib directory, so normal importing works.
+newContextInLib() {
+  var context = FakeExecutionContext();
+  context.selectedCallFrame = allowInterop(() => FakeCallFrame.inLib()
+    ..evaluationResult = FakeRemoteObject(value: 'foo/bar/web/main.html'));
+  return context;
+}
+
+/// A fake execution context. We write the fake in JS and access it using JS
+/// interop because we access JS methods that start with underscores, and if we
+/// create it in Dart those will get mangled and be harder to access.
+@JS('Dart.FakeExecutionContext')
+class FakeExecutionContext {
+  @JS('selectedCallFrame')
+  external set selectedCallFrame(x);
+}
+
+/// A fake that we use to replace the devtools Bindings namespace.
+class FakeChromeBindings {
+  get debuggerWorkspaceBinding => this;
+  UILocation _location = UILocation();
+  rawLocationToUILocation(_) => _location;
```

Comment:
why isn't this a getter called location?

Reply:
It's trying to mirror what the devtools Bindings does, and gets called from JS.

Reply:
sg